File: frontend/src/api/modules/tickets.ts
import type { RequestOptions } from '../request'
import request from '../request'

export type TicketMode = 'off' | 'manual' | 'auto'
export interface TicketPolicy { mode: TicketMode, targetLength: number | null }
export interface TicketSettings {
  enabled: boolean
  inject: boolean
  plusProLength: number
  businessLength: number
  defaultLength: number
  models: string[]
  intervalSeconds: number
  manualIntervalSeconds: number
  ttlSeconds: number
  refreshBeforeSeconds: number
  accounts: Record<string, TicketPolicy>
}
export interface TicketResult {
  accountId: string
  model: string
  httpStatus: number
  length: number
  matched: boolean
  checkedAt: number
  message: string
}
export interface TicketAccount {
  id: string
  name: string
  plan: string | null
  eligible: boolean
  targetLength: number
  policy: TicketPolicy
  models: { model: string, ready: boolean, expiresAt: number | null, lastResult: TicketResult | null, busy: boolean, retryAt: number | null, manualRetryAt: number | null }[]
}
export interface TicketLog {
  id: string
  accountName: string
  trigger: TicketMode
  proxyEndpoint: string
  proxyName: string
  targetLength: number
  startedAt: number
  durationMs: number
  retryAt: number | null
  result: TicketResult
}
export interface TicketProxyInput { id?: string, name: string, url?: string, savedProxyId?: string }
export interface TicketProxyView { id: string, name: string, endpoint: string, hasAuthentication: boolean }
export interface TicketPanel { revision: number, settings: TicketSettings, proxyConfigured: boolean, proxyCount: number, proxies: TicketProxyView[], accounts: TicketAccount[], logs: TicketLog[], logLimit: number }

export function getTicketPanel(options: RequestOptions = {}) {
  return request<TicketPanel>({
    url: '/api/admin/tickets',
    method: 'GET',
    ...options,
  })
}
export function saveTicketSettings(data: { revision: number, settings: TicketSettings, proxyUrl?: string, proxyPool?: string[], proxies?: TicketProxyInput[] }) {
  return request<TicketPanel>({
    url: '/api/admin/tickets',
    method: 'POST',
    data,
  })
}
export function probeTicket(data: { accountId: string, model: string }) {
  return request<TicketResult>({
    url: '/api/admin/tickets/probe',
    method: 'POST',
    data,
    timeout: 35000,
  })
}

File: backend/crates/gateway-admin/tests/model/tickets.rs
use gateway_admin::model::tickets::{TicketAccountPolicy, TicketMode, TicketSettings};

#[test]
fn ticket_plan_defaults_and_explicit_override_are_distinct() {
    let mut settings = TicketSettings::default();
    assert!(!settings.enabled);
    assert!(!settings.inject);
    assert_eq!(settings.manual_interval_seconds, 0);
    assert_eq!(settings.target_length("acct_a", Some("plus")), 292);
    assert_eq!(settings.target_length("acct_a", Some("pro")), 292);
    assert_eq!(settings.target_length("acct_a", Some("business")), 332);
    assert_eq!(settings.target_length("acct_a", Some("TEAM")), 332);
    settings.accounts.insert(
        "acct_a".into(),
        TicketAccountPolicy {
            mode: TicketMode::Manual,
            target_length: Some(312),
        },
    );
    assert_eq!(settings.target_length("acct_a", Some("pro")), 312);
    assert_eq!(settings.target_length("acct_b", Some("pro")), 292);
    assert!(settings.validate());
}

#[test]
fn ticket_settings_reject_unbounded_or_inconsistent_work() {
    let base = TicketSettings::default();
    for settings in [
        TicketSettings {
            interval_seconds: 0,
            ..base.clone()
        },
        TicketSettings {
            manual_interval_seconds: 86401,
            ..base.clone()
        },
        TicketSettings {
            ttl_seconds: 86400,
            ..base.clone()
        },
        TicketSettings {
            refresh_before_seconds: 3600,
            ..base.clone()
        },
        TicketSettings {
            plus_pro_length: 0,
            ..base.clone()
        },
        TicketSettings {
            models: vec![],
            ..base.clone()
        },
        TicketSettings {
            models: vec!["gpt-6-astra".into(), "gpt-6-astra".into()],
            ..base.clone()
        },
        TicketSettings {
            models: vec!["model\r\nheader".into()],
            ..base
        },
    ] {
        assert!(!settings.validate());
    }
}
